src/main/java/marquez/common/models/DatasourceUrn.java

```diff
@@ -1,24 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package marquez.common.models;
 
-import java.util.regex.Pattern;
+import static marquez.common.Preconditions.checkNotBlank;
+import static marquez.common.models.UrnPattern.URN_DELIM;
+import static marquez.common.models.UrnPattern.URN_PREFIX;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.StringJoiner;
 import lombok.EqualsAndHashCode;
+import lombok.Getter;
 import lombok.NonNull;
+import lombok.ToString;
+
+@EqualsAndHashCode
+@ToString
+public final class DatasourceUrn {
+  private static final int URN_PARTS = 2;
+  private static final UrnPattern URN_PATTERN = UrnPattern.of(UrnType.DATASOURCE, URN_PARTS);
 
-@EqualsAndHashCode(callSuper = false)
-public final class DatasourceUrn extends Urn {
-  private static final int NUM_COMPONENTS = 2;
-  private static final String URN_TYPE = "datasource";
-  private static final Pattern REGEX = buildPattern(URN_TYPE, NUM_COMPONENTS);
+  @Getter private final String value;
 
-  public DatasourceUrn(@NonNull String value) {
-    super(value, REGEX);
+  private DatasourceUrn(@NonNull final String value) {
+    URN_PATTERN.throwIfNoMatch(checkNotBlank(value));
+    this.value = value;
   }
 
-  public static DatasourceUrn from(@NonNull String type, @NonNull String name) {
-    final String value = fromComponents(URN_TYPE, type, name);
+  public static DatasourceUrn from(
+      @NonNull DatasourceType type, @NonNull DatasourceName datasourceName) {
+    final String value =
+        new StringJoiner(URN_DELIM)
+            .add(URN_PREFIX)
+            .add(UrnType.DATASOURCE.toString())
+            .add(type.toString())
+            .add(datasourceName.getValue())
+            .toString();
     return fromString(value);
   }
 
+  @JsonProperty
```

Comment:
Why is this necessary?

Reply:
Good catch. This actually should be `@JsonCreator`. That is, `fromString()` is the factory method Jackson should use on deserialization.